crates/admin/src/cluster_controller/logs_controller/nodeset_selection.rs

```diff
@@ -0,0 +1,356 @@
+        let candidates = WritableNodeSet::from(self.cluster_state, self.nodes_config);
+
+        let min_copies = replication_property.num_copies();
+        if candidates.len() < min_copies.into() {
+            trace!(
+                candidate_nodes_count = ?candidates.len(),
+                ?min_copies,
+                cluster_state = ?self.cluster_state,
+                nodes_config = ?self.nodes_config,
+                "Not enough writeable nodes to meet the minimum replication requirements"
+            );
+            return Err(NodeSelectionError::InsufficientWriteableNodes);
+        }
+
+        // ReplicationFactor(f+1) implies a minimum of 2f+1 nodes. At this point we are only
+        // calculating the nodeset floor size, the actual size will be determined by the specific
+        // strategy in use.
+        assert!(
+            min_copies < u8::MAX >> 1,
+            "The replication factor implies a cluster size that exceeds the maximum supported size"
+        );
+        let optimal_fault_tolerant_nodeset_size = (usize::from(min_copies) - 1) * 2 + 1;
+        assert!(
+            optimal_fault_tolerant_nodeset_size >= min_copies as usize,
+            "The calculated minimum nodeset size can not be less than the replication factor"
+        );
+
+        let (nodeset_min_size, nodeset_target_size) = match strategy {
+            NodeSetSelectionStrategy::StrictFaultTolerantGreedy => (
+                optimal_fault_tolerant_nodeset_size,
+                max(optimal_fault_tolerant_nodeset_size, candidates.len()),
+            ),
+        };
+
+        let nodeset = match strategy {
+            NodeSetSelectionStrategy::StrictFaultTolerantGreedy => {
+                let mut nodes = preferred_nodes
+                    .iter()
+                    .copied()
+                    .filter(|node_id| candidates.0.contains(node_id))
+                    .choose_multiple(rng, nodeset_target_size);
+
+                if nodes.len() < nodeset_target_size {
+                    let remaining = nodeset_target_size - nodes.len();
+                    nodes.extend(
+                        candidates
+                            .into_iter()
+                            .filter(|node_id| !preferred_nodes.contains(node_id))
+                            .choose_multiple(rng, remaining),
+                    );
+                }
+
+                let nodes_len = nodes.len();
+                let nodeset = NodeSet::from_iter(nodes);
+                assert_eq!(
+                    nodeset.len(),
+                    nodes_len,
+                    "We have accidentally chosen duplicate candidates during nodeset selection"
+                );
+                nodeset
+            }
+        };
+
+        // todo: implement location scope-aware selection
+        if nodeset.len() < nodeset_min_size {
+            trace!(
+                selected_nodes_count = ?nodeset.len(),
+                ?nodeset_target_size,
```

Comment:
Shouldn't we log `nodeset_min_size` here?

Reply:
Yes, this would have been very confusing! Wonder if it's still worth keeping the distinction actually. Anyway, now we will log the `nodeset_min_size`.